jobs/execute.ts

```diff
@@ -19,34 +30,61 @@ export async function runJob(name: JobName): Promise<boolean> {
-                        job_name: name,
-                        endedAt: { equals: null },
+        do {
+            try {
+                // Wait between 0 and 1000ms to reduce the likelihood of transaction deadlocks
+                // (as a lot of Cron Jobs fire at exactly the same time)
+                await new Promise((resolve) => setTimeout(resolve, Math.floor(Math.random() * 1000)));
+
+                jobRun = await prisma.$transaction(
+                    async (jobPrisma) => {
+                        const runningJob = await jobPrisma.job_run.findFirst({
+                            where: {
+                                job_name: name,
+                                endedAt: { equals: null },
+                            },
+                        });
+
+                        if (runningJob) {
+                            logger.warn(
+                                `Cannot concurrently execute Job '${name}' as it is already running on '${runningJob.worker}' since ${runningJob.startedAt}`
+                            );
+                            lockStatus = LockStatus.CONFLICT;
+                            return undefined;
+                        }
+
+                        lockStatus = LockStatus.AQUIRED;
+
+                        return await jobPrisma.job_run.create({
+                            data: { job_name: name, worker: process.env.DYNO ?? '?' },
+                        });
+                        // It is important that the transaction ends here and the INSERT above is commited
+                        // Otherwise we would continue execution, and the commit would be rolled back after the job actually executed
                     },
-                });
-
-                if (runningJob) {
-                    throw new Error(
-                        `Cannot concurrently execute Job '${name}' as it is already running on '${runningJob.worker}' since ${runningJob.startedAt}`
-                    );
-                }
-
-                return await jobPrisma.job_run.create({
-                    data: { job_name: name, worker: process.env.DYNO ?? '?' },
-                });
-
-                // It is important that the transaction ends here and the INSERT above is commited
-                // Otherwise we would continue execution, and the commit would be rolled back after the job actually executed
-            },
-            { isolationLevel: Prisma.TransactionIsolationLevel.Serializable }
-        );
+                    { isolationLevel: Prisma.TransactionIsolationLevel.Serializable }
+                );
+            } catch (error) {
+                // The transaction was aborted, likely because the DB rolled back the deadlock
+                lockStatus = LockStatus.ROLLBACK;
+                lockRetries -= 1;
+            }
+        } while (lockStatus === LockStatus.ROLLBACK && lockRetries > 0);
+
+        if (lockStatus === LockStatus.CONFLICT) {
+            return false;
```

Comment:
I think we should move the error from below to here
```        
logger.error(`Failure during Job Scheduling - This might leave the system in a locked state requiring manual cleanup!`, error);
```

Reply:
Yes, moved it into the ROLLBACK case, as that is the "not well defined" state. The CONFLICT state can happen if the same job is triggered multiple times - I'll also increase the logging to error so that we notice if there are many conflicts still